### ISSUE TITLE
Group Operation With Result

### ISSUE THREAD

**Issue by felix-dumit** [danthorpe/Operations#213]
Hi,

I'm refactoring part of my code into operations and I'm struggling with modelling a couple of scenarios. 
I'll post the 1st on here then the next on a separate issue.

```swift 
class CompressAssetOperation: GroupOperation, InjectionOperationType, ResultOperationType {
    var requirement:AssetConvertable?
    var result:CompressionResult?
    
    let asset:PHAsset
    var options:ImageCompressionOptions
    
    let loadDataOperation:LoadImageDataOperation
    let compressDataOperation:CompressImageDataOperation
    
    init(asset:PHAsset, options:ImageCompressionOptions) {
        self.asset = asset
        self.options = options

        loadDataOperation = LoadImageDataOperation(asset: asset)
        compressDataOperation = CompressImageDataOperation(options: options)
        compressDataOperation.injectResultFromDependency(loadDataOperation)
        
        super.init(operations:[loadDataOperation, compressDataOperation])
        
//        injectResultFromDependency(compressDataOperation) { op, compress, errors in
//            guard let requirement = compress.result else {
//                self.log.fatal("failed to get compressed data")
//                return
//            }
//            self.result = CompressionResult(compressedData: requirement, originalAsset: asset)
//            
//        }
    }
```

I have the operation as above, it needs to call the `LoadImageDataOperation` which produces the result that gets passed into the `CompressImageData` operation. 
The issue is that I want the `CompressAssetOperation` to have a result based on the compress operation, so I can then use this group operation and inject it somewhere else.

Basically what I need to do is the commented out part of my code, but if I uncomment it, none of the operations are actually executed.

Thanks for this library and any help.

**Comment by danthorpe** [comment]
Oh - this is interesting! Great question -  I haven't thought about using result injection inside a group. Let me have a think and get back to you.

**Comment by danthorpe** [comment]
Maybe I'm being silly here, but would this work?

```swift
class CompressAssetOperation: GroupOperation, InjectionOperationType, ResultOperationType {

    // etc 

    var result: CompressionResult? {
        return compress.result.map { CompressionResult(compressedData: $0, originalAsset: asset) }
    }

    // etc 
}
```

**Comment by felix-dumit** [comment]
I think that might work. Very simple :)
I probably don't need the `InjectionOperationType` anymore in this case right ?

**Comment by danthorpe** [comment]
Yup :+1: - you can remove those commented out lines in your sample. 

Also, it's probably best to move the checking of the result being nil (`self.log.fatal`) into the `execute` method of whatever operation will consume this result. I think instead, that operation should assert that it's requirement is fulfilled.

**Comment by felix-dumit** [comment]
When I'm trying to add this group operation as a parameter to another in `operation2.injectResultFromDependency(op)`, the compiler keeps complaining that I'm not passing the block to configure the dependency `{op, dep, errors in }`. 
So i've had to do this:
```swift
 op.injectResultFromDependency(dep) {op, dep, _ in  op.requirement = dep.result }
```

I wonder if it would be better for this operation to not be a group operation, and instead produce those child operations with `op.produceOperation()`

**Comment by danthorpe** [comment]
Hmm, I think you're on the right track with it being a Group - that makes more sense.

Does `operation2` conform to `AutomaticInjectionOperationType`?

**Comment by felix-dumit** [comment]
Oh I had it just as `InjectionOperationType`. My bad

**Comment by danthorpe** [comment]
That's no worries - this is great feedback for me. It's a bit subtle. The `InjectionResultType` is basically an empty protocol, but it provides the `injectResultFromDependency` API with the block. And `AutomaticInjectionOperationType` extends this and adds the `requirement` property, and then you can call the same API without the block.

My thinking behind this, is that it's quite possible that an `Operation` will have multiple independent dependencies, and so, `AutomaticInjectionOperationType` is too restrictive.

But for simple A -> B -> C type dependency chains it makes everything much easier. I think?

**Comment by felix-dumit** [comment]
The `AutomaticInjectionOperation` pretty much creates the same block as what I had to do manually so it makes sense

**Comment by danthorpe** [comment]
Yup. So it works?

**Comment by felix-dumit** [comment]
Yes it does work!

As I was typing out my 2nd question I came up with a solution, so I guess I'll have to bug you again some other time :)

**Comment by danthorpe** [comment]
Great! 

Yeah - feel free to create issues on here is you have questions, queries, bugs, or any problems really. I'll always try to help, and there are other contributors here that will often help out too :)

I'm gonna close this issue now if that's okay?

**Comment by felix-dumit** [comment]
Ok feel free to close it. 
Thanks!

**Comment by danthorpe** [comment]
:+1:
